File: referencia_temp/cronograma_manutencoes/scripts/mailer.py
"""
mailer.py — Envio de e-mail via SMTP interno CTG Brasil
Lê configurações de variáveis de ambiente para uso seguro no GitHub Actions.
"""

import os
import smtplib
import json
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.mime.base import MIMEBase
from email import encoders

# ─── CONFIG VIA ENV ───────────────────────────────────────────────────────────
SMTP_HOST  = os.environ.get("SMTP_HOST", "dtcvppostfix.ctgpar.ctgbr.com.br")
SMTP_PORT  = int(os.environ.get("SMTP_PORT", "25"))
REMETENTE  = os.environ.get("SMTP_FROM",  "cronograma_de_paradas@appmail.ctgbr.com.br")
SMTP_USER  = os.environ.get("SMTP_USER",  "")
SMTP_PASS  = os.environ.get("SMTP_PASS",  "")
USAR_TLS   = os.environ.get("SMTP_TLS",   "false").lower() == "true"

# Lista de destinatários: pode ser JSON inline ou caminho para arquivo .json
# Exemplo de JSON: [{"nome": "Fulano", "email": "fulano@ctgbr.com.br"}, ...]
_RECIPIENTS_ENV  = os.environ.get("RECIPIENTS", "")
_RECIPIENTS_FILE = os.environ.get("RECIPIENTS_FILE", "config/recipients.json")


def load_recipients() -> list[dict]:
    """Carrega a lista de destinatários do env ou do arquivo de config."""
    if _RECIPIENTS_ENV:
        try:
            return json.loads(_RECIPIENTS_ENV)
        except json.JSONDecodeError as e:
            raise ValueError(f"RECIPIENTS env inválido (JSON): {e}")
    if os.path.exists(_RECIPIENTS_FILE):
        with open(_RECIPIENTS_FILE, encoding="utf-8") as f:
            return json.load(f)
    raise FileNotFoundError(
        f"Nenhuma lista de destinatários encontrada. "
        f"Defina a env RECIPIENTS ou crie '{_RECIPIENTS_FILE}'."
    )


def enviar_email(
    destinatarios: list[str],
    assunto: str,
    mensagem: str,
    anexo_path: str | None = None,
    anexo_nome: str | None = None,
) -> None:
    """
    Envia e-mail HTML para uma lista de endereços.
    Opcionalmente anexa um arquivo (ex: o PDF do cronograma).
    """
    msg = MIMEMultipart()
    msg["From"]    = REMETENTE
    msg["To"]      = ", ".join(destinatarios)
    msg["Subject"] = assunto
    msg.attach(MIMEText(mensagem, "html"))

    # ── Anexo ──────────────────────────────────────────────────────────────
    if anexo_path and os.path.exists(anexo_path):
        with open(anexo_path, "rb") as f:
            part = MIMEBase("application", "octet-stream")
            part.set_payload(f.read())
        encoders.encode_base64(part)
        nome = anexo_nome or os.path.basename(anexo_path)
        part.add_header("Content-Disposition", f'attachment; filename="{nome}"')
        msg.attach(part)
    elif anexo_path:
        raise FileNotFoundError(f"Anexo não encontrado: {anexo_path}")

    # ── Envio ───────────────────────────────────────────────────────────────
    try:
        servidor = smtplib.SMTP(SMTP_HOST, SMTP_PORT, timeout=30)
        if USAR_TLS:
            servidor.starttls()
        if SMTP_USER and SMTP_PASS:
            servidor.login(SMTP_USER, SMTP_PASS)
        servidor.send_message(msg)
        servidor.quit()
        print(f"  E-mail enviado para: {', '.join(destinatarios)}")
    except Exception as e:
        raise RuntimeError(f"Falha no envio via SMTP ({SMTP_HOST}:{SMTP_PORT}): {e}") from e

File: backend/src/utils/mailer.js
/**
 * mailer.js — Envio de e-mail via SMTP
 * Suporta Gmail (recomendado) ou SMTP interno CTG Brasil
 * Lê configurações de variáveis de ambiente para uso seguro.
 */

import nodemailer from 'nodemailer';
import dotenv from 'dotenv';

dotenv.config();

// ─── CONFIG VIA ENV ───────────────────────────────────────────────────
const SMTP_HOST  = process.env.SMTP_HOST || 'dtcvppostfix.ctgpar.ctgbr.com.br';
const SMTP_PORT  = parseInt(process.env.SMTP_PORT || '25', 10);
const SMTP_FROM  = process.env.SMTP_FROM || 'ctg.engenharia@appmail.ctgbr.com.br';
const SMTP_USER  = process.env.SMTP_USER || '';
const SMTP_PASS  = process.env.SMTP_PASS || '';
const USAR_TLS   = process.env.SMTP_TLS === 'true';

// Gmail config (prioritário se configurado)
const GMAIL_USER = process.env.SMTP_GMAIL_USER || '';
const GMAIL_PASS = process.env.SMTP_GMAIL_PASS || '';

// ─── TRANSPORTER ─────────────────────────────────────────────────────
function createTransporter() {
  // GMAIL (prioritário)
  if (GMAIL_USER && GMAIL_PASS) {
    console.log('  📧 Usando Gmail:', GMAIL_USER);
    return nodemailer.createTransport({
      service: 'gmail',
      auth: {
        user: GMAIL_USER,
        pass: GMAIL_PASS, // App Password do Gmail
      },
    });
  }

  // SMTP interno CTG (fallback)
  console.log('  📧 Usando SMTP interno:', SMTP_HOST);
  const config = {
    host: SMTP_HOST,
    port: SMTP_PORT,
    secure: false, // false = tenta STARTTLS
    connectionTimeout: 30000,
    socketTimeout: 30000,
  };

  if (USAR_TLS) {
    config.requireTLS = true;
  }

  if (SMTP_USER && SMTP_PASS) {
    config.auth = {
      user: SMTP_USER,
      pass: SMTP_PASS,
    };
  }

  return nodemailer.createTransport(config);
}

// ─── ENVIAR EMAIL ─────────────────────────────────────────────────────
export async function enviarEmail({ destinatarios, assunto, mensagemHtml, anexoPath, anexoNome }) {
  const transporter = createTransporter();

  const mailOptions = {
    from: GMAIL_USER || SMTP_FROM, // Gmail usa o próprio e-mail como remetente
    to: Array.isArray(destinatarios) ? destinatarios.join(', ') : destinatarios,
    subject: assunto,
    html: mensagemHtml,
  };

  // Anexo opcional
  if (anexoPath) {
    mailOptions.attachments = [
      {
        filename: anexoNome || anexoPath.split('/').pop(),
        path: anexoPath,
      },
    ];
  }

  try {
    const info = await transporter.sendMail(mailOptions);
    console.log(`  ✅ E-mail enviado para: ${mailOptions.to} (${info.messageId})`);
    return info;
  } catch (err) {
    console.error('❌ Erro ao enviar e-mail:', err.message);
    throw new Error(`Falha no envio via SMTP: ${err.message}`);
  }
}

export default { enviarEmail };
